Fix FileResponse fallback code

mdellweg opened the pull request:
In case the FileResponse is using _sendfile_fallback and the requested
range is smaller then the chunk size, we need to only read and send
count bytes.

<!-- Thank you for your contribution! -->

## What do these changes do?

<!-- Please give a short brief about these changes. -->

## Are there changes in behavior for the user?

<!-- Outline any notable behaviour for the end users. -->

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->

## Checklist

- [ ] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

mdellweg committed:
Fix FileResponse fallback code

In case the FileResponse is using _sendfile_fallback and the requested
range is smaller then the chunk size, we need to only read and send
count bytes.

ipanova commented on CHANGES/6726.bugfix:
```diff
@@ -0,0 +1 @@
+Fixed a bug in ``FileResponses`` fallback code when the requested range is smaller than the chunk size.
```
s/FileResponses/FileResponse/

DavidRomanovizc commented:
Hi, @mdellweg, this issue is still relevant?

DavidRomanovizc commented:
@mdellweg, Since there has been no activity over the past year can you send an code example in which such a issue occurs. Also you need fill CONTRIBUTORS.txt.and change unit test.

Dreamsorcerer commented:
As mentioned, this still needs a regression test. It should go in test_sendfile_functional.py and use the `sender` fixture to show that the behaviour is the same with/without the fallback.

mdellweg commented:
I added a test.
But the weird thing is, I could not make the test fail. Even though I _think_ I found another place where this content length seems to be limited (weirdly so, after compression...) [0]. I even broke that (by commenting out line 126 [0]) and tracked the (now too large) chunk with a debugger to line 82 where it still wastoo large. But the test did not fail. Is it possible that the `client` used for testing here is more forgiving than it should be (for the purpose of testing anyway)?

How should I proceed?

[0] https://github.com/aio-libs/aiohttp/blob/master/aiohttp/http_writer.py#L126
[1] https://github.com/aio-libs/aiohttp/blob/master/aiohttp/http_writer.py#L82

Dreamsorcerer commented on tests/test_web_sendfile_functional.py:
```diff
@@ -1164,3 +1164,30 @@ async def handler(request: web.Request) -> web.FileResponse:
+    app.router.add_get("/", handler)
+    client = await aiohttp_client(app)
+
+    resp = await client.get("/", headers={"Range": "bytes=20-40"})
+    assert resp.status == 206
+    body = await resp.read()
```
It may also be important to use the context manager correctly if we want t test 2 requests in a row.
```suggestion
    async with client.get("/", headers={"Range": "bytes=20-40"}) as resp:
        assert resp.status == 206
        body = await resp.read()
```

Dreamsorcerer commented:
> If there's actually more data sent than the header says, then reading the _next_ response should produce an error. So, you could adapt the test to try sending and receiving 2 range requests in a row.

I should note that this error may only happen with the test client. If you try it outside of a test, then the client may silently retry the second request and succeed.

Dreamsorcerer commented:
Actually, looking at the code you've edited. Isn't it already doing that in the next couple of lines:
https://github.com/aio-libs/aiohttp/blob/237d46722a0e56363c4a07dcf31e73fe4c2ea392/aiohttp/web_fileresponse.py#L121

I think that code cuts it to the correct length. So, your min() is probably not making any difference functionally. It might save reading a little bit of data when it doesn't need it, but wastes a function call when it does, so I'm not clear if it would be more efficient or not.

mdellweg commented:
> Actually, looking at the code you've edited. Isn't it already doing that in the next couple of lines:

That is the line right after reading and handing down the too big chunk.
I think I don't quite understand what you intend with the next response.
Also i could no longer reproduce the bug in the environment i saw it originally.
As you say, now it's "just" the extra cost of the min call against reading a few more bytes than needed from the IO object.
I'd be fine with just closing this, given that i believe no extra bytes are actually transfered over the wire.

Dreamsorcerer commented:
> That is the line right after reading and handing down the too big chunk.

Ah, yes, it writes the existing chunk first. So, actually still seems like your fix might be relevant. Did you try the test with 2 range requests in a row?

mdellweg commented:
I'll have a shot.

mdellweg committed:
DO NOT MERGE

Trying to demonstrate that the test will not even fail when too many bytes are sent.

mdellweg commented:
Not sure if that's what you meant (looking at the test).
I also broke all the other places that I think prevent sending too many bytes, but still the test would not capture that. The client seems to comply to the content length more than to the tcp stream.

Dreamsorcerer commented:
> The client seems to comply to the content length more than to the tcp stream.

I think the client should definitely be hitting an error there if there's actually more data sent. Have you tried curl or any other clients and seen it sending more data?

bdraco commented:
All Python versions we support implement sendfile fallback internally now. So you would likely have to set the env var to reach the fallback code

bdraco commented:
Also we have been discussing dropping the aiohttp implementation now that asyncio does it for us

https://github.com/aio-libs/aiohttp/issues/10096

Dreamsorcerer commented:
I think from the last few comments, it sounds like this should just be closed now.
